kartothek/serialization/_parquet.py

```diff
@@ -95,85 +109,39 @@ def restore_dataframe(
-                            pd.RangeIndex(start=0, stop=parquet_file.metadata.num_rows)
-                        )
-                        df = pd.DataFrame(df, index=index)
-                        # convert back to table to keep downstream code untouched by this patch
-                        table = pa.Table.from_pandas(df)
-                    else:
-                        table = pq.read_pandas(reader, columns=columns)
-            finally:
-                reader.close()
-
-        if columns is not None:
-            missing_columns = set(columns) - set(table.schema.names)
-            if missing_columns:
-                raise ValueError(
-                    "Columns cannot be found in stored dataframe: {missing}".format(
-                        missing=", ".join(sorted(missing_columns))
-                    )
+                return _restore_dataframe(
+                    store=store,
+                    key=key,
+                    filter_query=filter_query,
+                    columns=columns,
+                    predicate_pushdown_to_io=predicate_pushdown_to_io,
+                    categories=categories,
+                    predicates=predicates,
+                    date_as_object=date_as_object,
                 )
-
-        table = _reset_dictionary_columns(table, exclude=categories)
-        df = table.to_pandas(categories=categories, date_as_object=date_as_object)
-        df.columns = df.columns.map(ensure_unicode_string_type)
-        if predicates:
-            df = filter_df_from_predicates(
-                df, predicates, strict_date_types=date_as_object
-            )
-        else:
-            df = filter_df(df, filter_query)
-        if columns is not None:
-            return df.reindex(columns=columns, copy=False)
-        else:
-            return df
+            except (AssertionError, OSError):
+                _logger.warning(
+                    msg=(
+                        f"Failed to restore dataframe, attempt {nb_retry} of {MAX_NB_RETRIES} with parameters "
+                        f"key: {key}, filter_query: {filter_query}, columns: {columns}, "
+                        f"predicate_pushdown_to_io: {predicate_pushdown_to_io}, categories: {categories}, "
+                        f"predicates: {predicates}, date_as_object: {date_as_object}."
+                    ),
+                    exc_info=True,
+                )
+                # we don't sleep when we're done with the last attempt
+                if nb_retry < (MAX_NB_RETRIES - 1):
+                    time.sleep(BACKOFF_TIME * 2 ** nb_retry)
+
+        raise ParquetReadError(
+            "Failed to restore dataframe after {MAX_NB_RETRIES} attempts. Parameters: "
+            f"key: {key}, filter_query: {filter_query}, columns: {columns}, "
+            f"predicate_pushdown_to_io: {predicate_pushdown_to_io}, categories: {categories}, "
+            f"predicates: {predicates}, date_as_object: {date_as_object}."
+        )
```

Comment:
please use 

```
raise ParquetReaderError(...) from original_excetion
```

to not loose that context information

----

I suggest to put the predicates last since they are the longest and may get lost if the exception message is truncated for whatever reason (e.g. distributed truncates at 1000chars)

Reply:
the problem is that the exception is out of scope at this point

Reply:
I guess I can work around it